### PR TITLE
Add option for a timeout while requesting a new connection

### PR DESCRIPTION
**Pull request by weiznich**
On windows the client goes in a infinite loop if it could not connect to the given server. This makes it possible to configure a timeout to prevent it.

### CODE REVIEW

**Comment by siddontang** [comment]
Is it better that using Timeout for all requests? Not only connect, but for read/write Timeout too.

File: http2/src/client_conf.rs [hunk 1]
```diff
@@ -1,6 +1,9 @@
+use std::time::Duration;
+
 #[derive(Default, Debug, Clone)]
 pub struct HttpClientConf {
     /// TCP_NODELAY
     pub no_delay: Option<bool>,
     pub thread_name: Option<String>,
+    pub timeout: Option<Duration>
```
Comment (stepancheg) [inline http2/src/client_conf.rs]: This should be called `connect_timeout`.

File: http2/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,7 @@ futures         = "0.*"
 #futures-cpupool = { git = "https://github.com/alexcrichton/futures-rs" }
 futures-cpupool = "0.*"
 tokio-core      = "0.*"
+tokio-timer     = "0.1"
```
Comment (stepancheg) [inline http2/Cargo.toml]: Please use `0.*` as in next lines.

File: http2/src/solicit/mod.rs [hunk 3]
```diff
@@ -404,6 +412,7 @@ impl Error for HttpError {
             HttpError::UnknownStreamId => "Attempted an operation with an unknown HTTP/2 stream ID",
             HttpError::UnableToConnect => "An error attempting to establish an HTTP/2 connection",
             HttpError::MalformedResponse => "The received response was malformed",
+            HttpError::Timeout => "Connection time out",
```
Comment (stepancheg) [inline http2/src/solicit/mod.rs]: Either enum value should be called `ConnectTimeout` (or smth like that), or message should say just "Timeout".

**Comment by stepancheg** [comment]
@siddontang 

> Is it better that using Timeout for all requests? Not only connect, but for read/write Timeout too.

IMO, read/write timeouts are not very useful practically. What required is per-call request timeout, deadline (instant) for for whole operation, not duration or single step.

**Comment by stepancheg** [comment]
@siddontang implementing request timeout is a different, harder issue, and connection timeout is useful itself, so I think I'll merge this patch as soon as you implement changes I requested.

**Comment by stepancheg** [comment]
Thanks!

**Pull request merged**